src/types/blockHeader.ts

```diff
@@ -79,3 +88,141 @@ export default interface BlockHeader {
    */
   spt: Map<number, Uint8Array>;
 }
+
+export const BLOCK_HEADER_SCHEMA = new NamedMapSchema([
+  {
+    key: 'fees',
+    valueSchema: new AddressSchema(),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'frac',
+    valueSchema: new Uint64Schema(),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'gen',
+    valueSchema: new StringSchema(),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'gh',
+    valueSchema: new FixedLengthByteArraySchema(32),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'prev',
+    valueSchema: new FixedLengthByteArraySchema(32),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'proto',
+    valueSchema: new StringSchema(),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'rate',
+    valueSchema: new Uint64Schema(),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'rnd',
+    valueSchema: new Uint64Schema(),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'rwcalr',
+    valueSchema: new Uint64Schema(),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'rwd',
+    valueSchema: new AddressSchema(),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'seed',
+    valueSchema: new FixedLengthByteArraySchema(32),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'ts',
+    valueSchema: new Uint64Schema(),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'txn',
+    valueSchema: new FixedLengthByteArraySchema(32),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'txn256',
+    valueSchema: new FixedLengthByteArraySchema(32),
+    required: true,
+    omitEmpty: true,
+  },
+  {
+    key: 'spt',
+    valueSchema: new UntypedSchema(), // TODO: fix
+    required: true,
+    omitEmpty: true,
+  },
+]);
+
+export function blockHeaderFromEncodingData(data: unknown): BlockHeader {
+  if (!(data instanceof Map)) {
+    throw new Error(`Invalid decoded BlockHeader: ${data}`);
+  }
+  return {
+    fees: data.get('fees'),
+    frac: data.get('frac'),
+    gen: data.get('gen'),
+    gh: data.get('gh'),
+    prev: data.get('prev'),
+    proto: data.get('proto'),
+    rate: data.get('rate'),
+    rnd: data.get('rnd'),
+    rwcalr: data.get('rwcalr'),
+    rwd: data.get('rwd'),
+    seed: data.get('seed'),
+    ts: data.get('ts'),
+    txn: data.get('txn'),
+    txn256: data.get('txn256'),
+    spt: data.get('spt'),
+  };
+}
+
+export function blockHeaderToEncodingData(
+  header: BlockHeader
+): Map<string, unknown> {
+  return new Map<string, unknown>([
+    ['fees', header.fees],
+    ['frac', header.frac],
+    ['gen', header.gen],
+    ['gh', header.gh],
+    ['prev', header.prev],
+    ['proto', header.proto],
+    ['rate', header.rate],
```

Comment:
missing `TxnCounter uint64 `codec:"tc"``
missing `RewardsLevel uint64 `codec:"earn"`` from rewardState, not sure what is the rule here.
also missing `partupdrmv` and all `UpgradeState` fields except `proto`

Reply:
Yes this block header representation is very outdated and wrong. We have an issue to fix in #846, and since this PR is large I didn't want to address that now. I only preserved the existing structure, even though it needs improvement